Update FIPS ingest docs for native Go FIPS 140-3 (#7818)
elastic/docs-content: deploy-manage/security/fips-ingest.md

Reviewer: spare key here - we need to delete these lines

```
applies_to:
  deployment: ga
```

Reviewer: this needs to be populated (I realize this was before your edits)

---
products:
  - id: apm
  - id: elastic-agent
  - id: beats
  - id: fleet
applies_to:
  stack: preview 9.1, ga 9.3+
---

# FIPS mode for Ingest tools [fips-ingest]

{{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} binaries are built and configured to use FIPS 140-2 compliant cryptography.

> 

Reviewer 2: I'm going to apply this so I can more easily take a look at this one.

---
products:
  - id: apm
  - id: elastic-agent
  - id: beats
  - id: fleet
applies_to:
  stack: preview 9.1, ga 9.3+
---

# FIPS mode for Ingest tools [fips-ingest]

{{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} binaries are built and configured to use FIPS-140-compliant cryptography:

* {applies_to}`stack: preview 9.1-9.2` is configured to use FIPS 140-2 compliant cryptography.

> * {applies_to}`stack: ga 9.3+` is configured to use FIPS 140-3 compliant cryptography using Go's native FIPS 140-3 module (`GOFIPS140=v1.0.0`, CMVP Certificate [#5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247)). Specific components and configurations that fall outside the certified boundary are documented in the [Limitations](#ingest-limitations-all) section.

Reviewer 2: Is this really 9.3+ for Elastic Agent? I only see this in the 9.5 release notes and it doesn't look like https://github.com/elastic/elastic-agent/pull/14953 was backported.

Reviewer 2: I see this was brought up in https://github.com/elastic/docs-content/pull/7818#discussion_r3759764624. But I don't think the suggestion matches the real applicability.

- For Filebeat, Metricbeat, and APM Server, it looks ike native Go FIPS was backported to 9.3 and 9.4
- For agent and fleet, i only see 9.5

> * {applies_to}`stack: ga 9.3+` is configured to use FIPS 140-3 compliant cryptography using Go's native FIPS 140-3 module (`GOFIPS140=v1.0.0`, CMVP Certificate [#5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247)). Specific components and configurations that fall outside the certified boundary are documented in the [Limitations](#ingest-limitations-all) section.

Reviewer 2: Would it be more accurate to say this?

```suggestion
* {applies_to}`stack: preview 9.1-9.2` FIPS binaries use FIPS 140-2 compliant cryptography.
* {applies_to}`stack: ga 9.3+` {{filebeat}}, {{metricbeat}}, and {{apm-server}} FIPS binaries use FIPS 140-3 compliant cryptography using Go's native FIPS 140-3 module (`GOFIPS140=v1.0.0`, CMVP Certificate [#5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247)).
* {applies_to}`stack: ga 9.5+` {{agent}} and {{fleet-server}} FIPS binaries use the same FIPS 140-3 module. On 9.3 and 9.4, {{agent}} and {{fleet-server}} FIPS binaries continue to use FIPS 140-2 cryptography.

Specific components and configurations that fall outside the certified boundary are documented in the [Limitations](#ingest-limitations-all) section.
```

Author: yes, it matches reality more than the previous version. I apparently mixed beats and elastic agents

# FIPS mode for Ingest tools [fips-ingest]

{{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} binaries are built and configured to use FIPS 140-2 compliant cryptography.

> {applies_to}`stack: ga 9.3+` {{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} binaries are built and configured to use FIPS 140-3 compliant cryptography via Go's native FIPS 140-3 module (`GOFIPS140=v1.0.0`, CMVP Certificate [#5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247)). This applies to patch versions 9.5.0, 9.4.4, and 9.3.8 and later. Specific components and configurations that fall outside the certified boundary are documented in the [Limitations](#ingest-limitations-all) section.

Author: @shainaraskas @nastasha-solomon Is it correct usage of `{applies_to}` if we are referring to 9.5.0+, 9.4.4+, and 9.3.8+ ?

# FIPS mode for Ingest tools [fips-ingest]

> {{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} are FIPS 140-3 capable. They use Go's native FIPS 140-3 module (`GOFIPS140=v1.0.0`, CMVP Certificate [#5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247)) for cryptographic operations. Specific components and configurations that fall outside the certified boundary are documented in the limitations sections below.

Reviewer 3: It's been a while since I've seen docs about FIPs, so take my suggestions with a grain of salt. From what I recall, FIPs-capable services or applications are slightly different from FIPs-compatible ones: 

- FIPS capable: Underlying architecture supports FIPS encryption standards, but the service/app is not federally certified.
- FIPS compatible: Service/app can communicate with FIPS systems but might not be using FIPS encryption.

I've also seen these terms used interchangeably, so this might be a nitpicky observation on my part. Just wanted to raise this in case it really is something that needs to be consistent across these docs.

Author: I think it is very grey area and I took stricter/safer approach here, but the true is that our fips compliance depends a lot on configuration and set of the components being used. FIPS compatible is safer statement, but I believe FIPS capability can be also defended. The situation is not changing much between the previous and current builds, so I can move the narration more into FIPS capable direction

Author: I think I misread the intention of your comment/mixed with the other one, and after revisiting my change i think  I will go back to previous wording

## FIPS-compatible binaries and configuration [fips-binaries]

FIPS compatible binaries for {{agent}}, {{fleet}}, {{filebeat}}, {{metricbeat}}, and {{apm-server}} are available for [download](https://www.elastic.co/downloads). Look for the `Linux 64-bit (FIPS)` or `Linux aarch64 (FIPS)` platform option on the product download pages for {{agent}} and {{fleet}}, {{filebeat}}, and {{metricbeat}}. Look for the `Linux x86_64 (FIPS)` or `Linux aarch64 (FIPS)` platform option on the {{apm-server}} download page.

:::{important}
The default configurations provided in the binaries are designed for FIPS-capable operation. Review the limitations below to ensure your full deployment stays within the certified boundary.
:::

## Limitations [ingest-limitations-all]

### TLS [ingest-limitations-tls]

> Only FIPS 140-3 compliant TLS protocols, ciphers, and curve types are allowed to be used as listed in the following section.

Reviewer 3: Continuing [this comment](https://github.com/elastic/docs-content/pull/7818/changes#r3730865590), FIPS-compliant means something different as well.

Author: I agree, but regarding the protocols, ciphers and curve types, they are well defined in the FIPS and this section is more like summary than claiming anything about our code/config

* The supported TLS versions are `TLS v1.2` and `TLS v1.3`.
* The supported cipher suites are:
  * `TLS v1.2`: `ECDHE-RSA-AES-128-GCM-SHA256`, `ECDHE-RSA-AES-256-GCM-SHA384`, `ECDHE-ECDSA-AES-128-GCM-SHA256`, `ECDHE-ECDSA-AES-256-GCM-SHA384`
  * `TLS v1.3`: `TLS-AES-128-GCM-SHA256`, `TLS-AES-256-GCM-SHA384`
* The supported curve types are `P-256`, `P-384` and `P-521`.

> * The minimum key length is 2048 bits for RSA keys. EC key size is determined by the curve in use; see the supported curve types above.

Reviewer 3: If you can link the docs that readers should refer to, that'd improve the UX here.

```suggestion
* The minimum key length is 2048 bits for RSA keys. EC key size is determined by the curve in use. Refer to the supported curve types above.
```

Support for encrypted private keys is not available, as the cryptographic modules used for decrypting password protected keys are not FIPS validated. If an output or any other component with an SSL key that is password protected is configured, the components will fail to load the key. When running in FIPS mode, you must provide non-encrypted keys.
Be sure to enforce security in your FIPS environments through other means, such as strict file permissions and access controls on the key file itself, for example.

> These TLS related restrictions apply to all components listed in the preceding section.

Reviewer 3: Same comment here about linking directly to the relevant docs instead of using directional language. (Only if possible ofc)

### General output and input limitations (Kerberos protocol) [ingest-inputoutput-limitations]

> The Kerberos protocol is not supported for outputs and inputs in Filebeat, Metricbeat, APM Server, and Fleet Server configurations. This also restricts the available `sasl.mechanism` for the Kafka output, where only `PLAIN` is supported.

Reviewer 3: ```suggestion
The Kerberos protocol is not supported for outputs and inputs in {{filebeat}}, {{metricbeat}}, {{apm-server}}, and {{fleet-server}} configurations. This also restricts the available `sasl.mechanism` for the Kafka output, where only `PLAIN` is supported.
```

### TLS [ingest-limitations-tls]

Only FIPS-compliant TLS protocols, ciphers, and curve types are allowed, as defined by [CMVP Certificate #5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247):
* The supported TLS versions are `TLS v1.2` and `TLS v1.3`.
* The supported cipher suites are:
  * `TLS v1.2`: `ECDHE-RSA-AES-128-GCM-SHA256`, `ECDHE-RSA-AES-256-GCM-SHA384`, `ECDHE-ECDSA-AES-128-GCM-SHA256`, `ECDHE-ECDSA-AES-256-GCM-SHA384`
  * `TLS v1.3`: `TLS-AES-128-GCM-SHA256`, `TLS-AES-256-GCM-SHA384`
* The supported curve types are `P-256`, `P-384` and `P-521`.

> * {applies_to}`stack: ga 9.3+` The minimum key length is 2048 bits for RSA keys. EC key size is determined by the curve in use.

Reviewer 2: Is this really 9.3+? It looks like this requirement has been around for a lot longer: https://github.com/elastic/elastic-agent/pull/7912

Author: I re-validated it and it is true. It comes wit FIPS 140-2 as well

### TLS [ingest-limitations-tls]

Only FIPS-compliant TLS protocols, ciphers, and curve types are allowed, as defined by [CMVP Certificate #5247](https://csrc.nist.gov/projects/cryptographic-module-validation-program/certificate/5247):
* The supported TLS versions are `TLS v1.2` and `TLS v1.3`.
* The supported cipher suites are:
  * `TLS v1.2`: `ECDHE-RSA-AES-128-GCM-SHA256`, `ECDHE-RSA-AES-256-GCM-SHA384`, `ECDHE-ECDSA-AES-128-GCM-SHA256`, `ECDHE-ECDSA-AES-256-GCM-SHA384`
  * `TLS v1.3`: `TLS-AES-128-GCM-SHA256`, `TLS-AES-256-GCM-SHA384`
* The supported curve types are `P-256`, `P-384` and `P-521`.

> * {applies_to}`stack: ga 9.3+` The minimum key length is 2048 bits for RSA keys. EC key size is determined by the curve in use.

Author: ```suggestion
* The minimum key length is 2048 bits for RSA keys. EC key size is determined by the curve in use.
```

* Agent upgrade artifact verification uses GPG signature checking, which is outside the FIPS module boundary.

### {{agent}} in OpenTelemetry mode (EDOT) [ingest-limitations-edot]
```{applies_to}
stack: ga 9.3+
```

When running {{agent}} in [OpenTelemetry mode](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/otel/README.md) (EDOT), these additional limitations apply.

* **Azure integrations**: PKCS#12 (`.pfx`) client certificates are not supported for Azure Active Directory authentication. Use client secrets, workload identity, managed identity, or PEM-encoded certificate and key files instead.
* **Kafka metrics receiver**: SASL GSSAPI (Kerberos) is not supported. SASL SCRAM (SCRAM-SHA-256, SCRAM-SHA-512) is not FIPS-compliant. Use SASL/PLAIN over TLS or mTLS.
* The following receivers are not supported in FIPS mode:
  * **MongoDB receiver**
  * **MySQL receiver**

>   * **Microsoft SQL Server receiver**

Reviewer 2: Can we add in Prometheus and Kafka here since they're mentioned above in the 9.1-9.2 bullet

```suggestion
* The following components are not included in FIPS builds:
  * **Prometheus receiver**
  * **Kafka receiver** and **Kafka exporter**
* The following receivers are not supported in FIPS mode:
  * **MongoDB receiver**
  * **MySQL receiver**
  * **Microsoft SQL Server receiver**
```

* The [Fingerprint processor](beats://reference/metricbeat/fingerprint.md) does not support the md5 and sha1 method.
* The [Community ID Network Flowhash processor](beats://reference/metricbeat/community-id.md) is not supported.
* The [Azure module](beats://reference/metricbeat/metricbeat-module-azure.md) is not supported. The [Add Cloud Metadata processor](beats://reference/metricbeat/add-cloud-metadata.md) does not support the Azure Virtual Machine provider.
* The [Google Cloud Platform module](https://www.elastic.co/docs/reference/beats/metricbeat/metricbeat-module-gcp) is not supported.
* The [Beta KVM module](beats://reference/metricbeat/metricbeat-module-kvm.md) is not supported.
* The [Mongo DB module](beats://reference/metricbeat/metricbeat-module-mongodb.md) is not supported.
* The [MySQL](beats://reference/metricbeat/metricbeat-module-mysql.md), [PostgreSQL](beats://reference/metricbeat/metricbeat-module-postgresql.md), [MSSQL](beats://reference/metricbeat/metricbeat-module-mssql.md) and [SQL](beats://reference/metricbeat/metricbeat-module-sql.md) modules are not supported.
* The [Oracle module](beats://reference/metricbeat/metricbeat-module-oracle.md) is not supported.

> ### Elastic Agent and Fleet Server [ingest-limitations-agent]

Reviewer 3: ```suggestion
### {{agent}} and {{fleet-server}} [ingest-limitations-agent]
```

When you use {{agent}} and {{fleet-server}}, these limitations apply:
* Some Elastic Integrations are not FIPS compatible, as they depend on functionality that is not supported for FIPS configuration. In general, when using {{agent}} and {{fleet-server}}, the same restrictions listed previously for {{metricbeat}} and {{filebeat}} modules, inputs, and processors apply.
* Agent upgrade artifact verification uses GPG signature checking, which is outside Go's certified FIPS module boundary.

> ### Elastic Agent in OpenTelemetry mode (EDOT) [ingest-limitations-edot]

Reviewer 3: ```suggestion
### {{agent}} in OpenTelemetry mode (EDOT) [ingest-limitations-edot]
```

When running {{agent}} in [OpenTelemetry mode](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/otel/README.md) (EDOT), these additional limitations apply:

* **Azure integrations**: PKCS#12 (`.pfx`) client certificates are not supported for Azure Active Directory authentication. Use client secrets, workload identity, managed identity, or PEM-encoded certificate and key files instead.
* **Kafka metrics receiver**: SASL GSSAPI (Kerberos) is not supported. SASL SCRAM (SCRAM-SHA-256, SCRAM-SHA-512) is not FIPS-compliant. Use SASL/PLAIN over TLS or mTLS.
* **MongoDB receiver**: Not supported in FIPS mode.
* **MySQL receiver**: Not supported in FIPS mode.
* **Microsoft SQL Server receiver**: Not supported in FIPS mode.

> * **Kerberos authentication for beat receivers**: Kerberos/GSSAPI authentication for beat receivers (Filebeat receiver, Metricbeat receiver) is not supported.

Reviewer 3: ```suggestion
* **Kerberos authentication for beat receivers**: Kerberos/GSSAPI authentication for beat receivers ({{filebeat}} receiver, {{metricbeat}} receiver) is not supported.
```

* **API key authentication extension**: Not FIPS-compliant. Uses a PBKDF2 implementation outside the certified FIPS module boundary when active.

If you are using a component not listed here and are unsure whether it is FIPS compliant, contact [Elastic Support](https://www.elastic.co/support).

### Elastic Integrations that are not FIPS compatible [ingest-limitations-integrations]

> The following Elastic Integrations (Fleet-managed) use cryptographic implementations outside Go's certified FIPS module boundary for core functionality and **cannot** be used in FIPS environments. For EDOT-specific limitations, see [Elastic Agent in OpenTelemetry mode (EDOT)](#ingest-limitations-edot).

Reviewer 3: ```suggestion
The following Elastic Integrations (Fleet-managed) use cryptographic implementations outside Go's certified FIPS module boundary for core functionality and **cannot** be used in FIPS environments. For EDOT-specific limitations, refer to [Elastic Agent in OpenTelemetry mode (EDOT)](#ingest-limitations-edot).
```
